README.md
# Data-efficient Reinforcement Learning in Latent Spaces for Robotic Clothing Assistance
---

Paper submitted to NIPS 2016 Workshop on Practical Bayesian Nonparametrics.


Reviews/Changes
# Changes to NIPS paper
---

* Provide formulation for DMP and relationship with latent space embedding (Done)
* Provide explanation on inference of ARD weight parameters (Done)
* Improve explanation of BGPLVM inference algorithm (Done)
* Emphasis on variability of dataset which avoids overfitting (Done)
* Remove second instance of video link and check typesetting of video link url (Done)
* Include explanation on computation cost of BGPLVM in results section (Done)
* Improve explanation for rl curves figure on why GPLVM performs worse than PCA (Done)
* Update equation 5 on reward function to L2 norm (Done)
* Remove the term furthermore in final paragraph of introduction section. (Done)
* Provide explanation of rollout in Results section (Done)
* Fix all minor typos indicated by reviewer 1 (Done)
* Correct marginal likelihood term to conditional likelihood (Done)
* Change wording from latent manifold to latent space (Done)
* Change sentence structure on generalization ability of work by Tamei (Done)


Experiments/viapointExtract.py
#!/usr/bin/env python

# viapoint_extract.py: python class implementation to extract viapoints
# Author: Nishanth Koganti
# Date: 2016/10/27

# import modules
import argparse
import matplotlib
import numpy as np
import cPickle as pickle
import matplotlib.pyplot as plt

# matplotlib default settings
font = {'family': 'sans-serif',
        'weight': 'bold',
        'size':   10}

matplotlib.rc('font', **font)

class ViaPoint:
    """class implementation for via-point extraction."""
    def __init__(self, data, nDims = 1, nViaPoints=6, errThresh=0.05, viaThresh=0.01):
        """initialization function with variable setting."""
        # setting the viapoint parameters
        self.params = {}
        self.nDims = nDims
        self.tDat = data[:, 0]
        self.data = data[:, 1:]
        self.errThresh = errThresh
        self.viaThresh = viaThresh
        self.nViaPoints = nViaPoints
        self.nPoints = self.tDat.shape[0]


    def compute(self):
        """function to compute via-points which is main interface to class."""
        # initialization
        out = np.zeros(self.data.shape)

        # loop over the dimensions
        for j in range(self.nDims):
            # via-Point algo doesn't work for small variations in angle
            if (self.data[:, j]).std() > self.errThresh:
                # smoothing trajectory
                out[:, j], param = self.viapoints(self.data[:, j], self.nViaPoints)
                self.params[j] = param
            else:
                # if angle doesn't change much then via-point not applied
                out[:, j] = self.data[:, j].copy()

        # output data
        out = np.concatenate((np.transpose(np.atleast_2d(self.tDat)),out), axis=1)
        return out, self.params

    def generate(self, params):
        """function to interpolate for given viapoints."""
        # initialization
        out = self.data.copy()

        # loop over the joints
        for j in params.keys():
            out[:, j] = self.interpolate(params[j])

        out = np.concatenate((np.transpose(np.atleast_2d(self.tDat)),out), axis=1)
        return out

    # function to interpolate between via-points
    def interpolate(self, vpoints):
        """function to interpolate between given viapoints."""
        # number of points for interpolation
        nPoints = self.tDat.shape[0]

        # create output array
        oDat = np.zeros(nPoints)

        # looping over pairs of via points
        for via in range(vpoints.shape[0]-1):
            # initial conditions
            t0 = vpoints[via, 0]
            y0 = vpoints[via, 1]
            yd0 = vpoints[via, 2]
            ind0 = np.int(vpoints[via, 3])

            # final conditions
            t1 = vpoints[via+1, 0]
            y1 = vpoints[via+1, 1]
            yd1 = vpoints[via+1, 2]
            ind1 = np.int(vpoints[via+1, 3])

            for ind in range(ind0, ind1+1):
                t = self.tDat[ind]
                oDat[ind] = y0+yd0*(t-t0)+((20.0*y1-20.0*y0-(8.0*yd1+12.0*yd0)*(t1-t0))/(2.0*(t1-t0)**3.0))*(t-t0)**3.0+(((30.0*y0-30.0*y1)+(14.0*yd1+16.0*yd0)*(t1-t0))/(2.0*(t1-t0)**4.0))*(t-t0)**4.0+((12.0*y1-12.0*y0-(6.0*yd1+6.0*yd0)*(t1-t0))/(2.0*(t1-t0)**5.0))*(t-t0)**5.0

        return oDat

    # function for extracting via points
    def viapoints(self, yDat, nVia):
        """function to extract viapoints for given trajectory."""
        # initialization
        nPoints = yDat.shape[0]

        # sampling period
        ts = self.tDat[-1]/nPoints

        # setting first and last via points
        vpoints = np.zeros((2, 4))
        vpoints[0, :] = np.asarray([self.tDat[0], yDat[0], 0.0, 0])
        vpoints[-1, :] = np.asarray([self.tDat[-1], yDat[-1], 0.0, nPoints-1])

        # loop over the via points
        oDat = np.zeros(nPoints)
        for runs in range(nVia-2):
            oDat = self.interpolate(vpoints)

            errors = np.sqrt((yDat-oDat)**2)

            mI = errors.argmax()
            vpoint = np.asarray([self.tDat[mI], yDat[mI],
                                 (yDat[mI+2]-yDat[mI-2])/(4*ts), mI])
            vpoints = np.concatenate((vpoints, np.atleast_2d(vpoint)), axis=0)
            vpoints = vpoints[np.argsort(vpoints[:, 0]), :]

            meanError = errors.mean()
            if meanError < self.viaThresh:
                break

        oDat = self.interpolate(vpoints)

        return oDat, vpoints

    def plot(self, params):
        """funtion to plot the computed results."""
        # plot initialization
        plt.figure()
        nPlots = len(params.keys())

        # looping over viapoints
        for i,j in enumerate(params.keys()):
            # via points extraction
            param = params[j]

            raw = self.data[:,j]
            smooth = self.interpolate(param)

            # plot figures
            plt.subplot(nPlots*100+11+i)

            plt.plot(self.tDat, raw, '-k', linewidth=2, label='raw')
            plt.plot(self.tDat, smooth, '-b', linewidth=2, label='smooth')
            plt.plot(param[:, 0], param[:, 1], 'ob', markersize=15, label='viapoints')

            plt.xlabel('Time', fontsize=12, fontweight='bold')
            plt.ylabel('Traj', fontsize=12, fontweight='bold')
            plt.title('Index %d' % (j+1), fontsize=15, fontweight='bold')
            if i == 0:
                plt.legend()

        plt.tight_layout()
        plt.show()


Experiments/plotFuncs.py
#!/usr/bin/env python

# plotFuncs.py: plot functions for data inspection
# Author: Nishanth Koganti
# Date: 2016/10/27

import sys
import GPy
import pydmps
import matplotlib
import numpy as np
from matplotlib import cm
from matplotlib import pyplot as plt
from GPy.plotting.matplot_dep.controllers.imshow_controller import ImshowController
from GPy.inference.latent_function_inference import InferenceMethodList, VarDTC, VarDTC_minibatch

def plotTraj(Dataset, plotType = 0, jointIndex = np.arange(7), labels = ['Train','Test'], colors=['b','r']):
    """function to plot multiple joint tracks."""
    timeData = []
    leftData = []
    rightData = []

    LEFT_ANGLE_OFFSET = 1
    RIGHT_ANGLE_OFFSET = 8

    # loop over first plotNum files
    for data in Dataset.values():
        timeData.append(data[:, 0])
        leftData.append(data[:, LEFT_ANGLE_OFFSET+jointIndex])
        rightData.append(data[:, RIGHT_ANGLE_OFFSET+jointIndex])

    jointData = [leftData, rightData]

    # number of joints to plot
    xlabel = 'Time(sec)'
    arms = ['Left', 'Right']
    nJoints = jointIndex.size
    if plotType == 0:
        ylabels = 7*['Joint Angle (rad)']
    else:
        ylabels = 3*['Position (m)']+4*['Angle (rad)']

    # plot all the joint data
    for ind in range(2):
        fig = plt.figure(figsize=(10, 2*nJoints))
        for i, jI in enumerate(jointIndex):
            plt.subplot(nJoints, 1, i+1)

            # plot all the tracks
            for n in range(len(Dataset.values())):
                timeDat = timeData[n]
                nSamples = jointData[ind][n].shape[0]
                plt.plot(timeDat, jointData[ind][n][:, i], label=labels[n], color=colors[n], linewidth=2)

            plt.xlabel(xlabel, fontsize=12, fontweight='bold')
            plt.ylabel(ylabels[i], fontsize=12, fontweight='bold')

            if plotType == 0:
                plt.title('%s Joint %d' % (arms[ind], jI+1), fontsize=15, fontweight='bold')
            else:
                plt.title('%s Pose %d' % (arms[ind], jI+1), fontsize=15, fontweight='bold')

            # plot legend only for 1st sub plot
            if i == 0:
                plt.legend()

        # adjust subplots for legend
        fig.subplots_adjust(top=0.96, right=0.8)
        plt.tight_layout()

    # show all the plots
    plt.show()

def plotTraj2(Dataset, points = None, colors={'Train':'b','Test':'r'}):
    """function to plot multiple joint tracks."""
    timeData = {}
    latentData = {}

    # loop over first plotNum files
    for key,data in Dataset.iteritems():
        timeData[key] = data[:, 0]
        latentData[key] = data[:, 1:]

    # number of latent dims to plot
    xlabel = 'Time(sec)'
    nDim = latentData[key].shape[1]
    ylabels = nDim*['Latent Position']

    # plot all the latent data
    fig = plt.figure(figsize=(10, 2*nDim))
    for i in range(nDim):
        plt.subplot(nDim, 1, i+1)

        # plot all the tracks
        for n,key in enumerate(Dataset.keys()):
            timeDat = timeData[key]
            nSamples = latentData[key].shape[0]
            plt.plot(timeDat, latentData[key][:, i], label=key, color=colors[key], linewidth=2)

        if points:
            plt.plot(points[i][:, 0], points[i][:, 1], 'ob', markersize=15, label='viapoints')

        plt.xlabel(xlabel, fontsize=12, fontweight='bold')
        plt.ylabel(ylabels[i], fontsize=12, fontweight='bold')
        plt.title('Dim %d' % (i+1), fontsize=15, fontweight='bold')

    # adjust subplots for legend
    plt.tight_layout()

    # show all the plots
    plt.show()

def pcaLatent(model, trainInput, testInput):
    # get latent space plot parameters
    testData = model.transform(testInput)
    trainData = model.transform(trainInput)
    return trainData, testData

def gplvmLatent(model, trainInput, testInput):
    # get latent space plot parameters
    qDim = model.X.shape[1]
    testData = np.zeros((testInput.shape[0], qDim))
    trainData = np.zeros((trainInput.shape[0], qDim))

    for n in range(trainInput.shape[0]):
        # infer latent position
        xTrain, _ = model.infer_newX(np.atleast_2d(trainInput[n,:]), optimize=True)

        # update parameter
        trainData[n,:] = xTrain
        sys.stdout.write('.')
    sys.stdout.write('\n')

    for n in range(testInput.shape[0]):
        # infer latent position
        xTest, _ = model.infer_newX(np.atleast_2d(testInput[n,:]), optimize=True)

        # update parameter
        testData[n,:] = xTest
        sys.stdout.write('.')
    sys.stdout.write('\n')

    return trainData, testData

def plotLatent(model, trainInput, testInput, nPoints=400, wThresh=0.05):
    sTest = 200
    sTrain = 150
    resolution = 50

    testMarker = 's'
    trainMarker = 'o'

    testLabels = [(1,0,0)]*nPoints
    trainLabels = [(0,0,1)]*nPoints

    # get active dimensions
    scales = model.kern.input_sensitivity(summarize=False)
    scales = scales/scales.max()
    activeDims = np.where(scales > wThresh)

    fig = plt.figure()
    ax = fig.add_subplot(111)

    # get latent space plot parameters
    plotIndices = [0,1]
    qDim = model.X.mean.shape[1]
    input1, input2 = plotIndices

    # loop over test data
    saveDim = activeDims[0].shape[0]
    testData = np.zeros((testInput.shape[0], saveDim))
    trainData = np.zeros((trainInput.shape[0], saveDim))

    for n in range(trainInput.shape[0]):
        # infer latent position
        xTrain, _ = model.infer_newX(np.atleast_2d(trainInput[n,:]), optimize=True)

        # update parameter
        trainData[n,:] = xTrain.mean[0,activeDims[0]]
        sys.stdout.write('.')
    sys.stdout.write('\n')

    for n in range(testInput.shape[0]):
        # infer latent position
        xTest, _ = model.infer_newX(np.atleast_2d(testInput[n,:]), optimize=True)

        # update parameter
        testData[n,:] = xTest.mean[0,activeDims[0]]
        sys.stdout.write('.')
    sys.stdout.write('\n')

    # compute plot limits
    xmin, ymin = trainData[:, [input1, input2]].min(0)
    xmax, ymax = trainData[:, [input1, input2]].max(0)
    x_r, y_r = xmax-xmin, ymax-ymin
    xmin -= .1*x_r
    xmax += .1*x_r
    ymin -= .1*y_r
    ymax += .1*y_r

    # plot the variance for the model
    def plotFunction(x):
        Xtest_full = np.zeros((x.shape[0], qDim))
        Xtest_full[:, [input1, input2]] = x
        _, var = model.predict(np.atleast_2d(Xtest_full))
        var = var[:, :1]
        return -np.log(var)

    x, y = np.mgrid[xmin:xmax:1j*resolution, ymin:ymax:1j*resolution]
    gridData = np.hstack((x.flatten()[:, None], y.flatten()[:, None]))
    gridVariance = (plotFunction(gridData)).reshape((resolution, resolution))

    varianceHandle = plt.imshow(gridVariance.T, interpolation='bilinear', origin='lower', cmap=cm.gray,
                                extent=(xmin, xmax, ymin, ymax))

    # test and training plotting
    testHandle = ax.scatter(testData[:, input1], testData[:, input2], marker=testMarker, s=sTest, c=testLabels,
                            linewidth=.2, edgecolor='k', alpha=1.)
    trainHandle = ax.scatter(trainData[:, input1], trainData[:, input2], marker=trainMarker, s=sTrain, c=trainLabels,
                            linewidth=.2, edgecolor='k', alpha=1.)

    ax.grid(b=False)
    ax.set_aspect('auto')
    ax.tick_params(axis='both', labelsize=20)
    ax.set_xlabel('Latent Dimension %i' % (input1+1), fontsize=25, fontweight='bold')
    ax.set_ylabel('Latent Dimension %i' % (input2+1), fontsize=25, fontweight='bold')

    ax.set_xlim((xmin, xmax))
    ax.set_ylim((ymin, ymax))

    properties = {'weight':'bold','size':25}
    plt.legend([trainHandle, testHandle], ['Train', 'Test'], prop=properties)

    fig.canvas.draw()
    fig.tight_layout()
    fig.canvas.draw()
    plt.show()

    return trainData, testData
